lib/http3/server.c

```diff
@@ -1452,6 +1452,7 @@ static void establish_tunnel(h2o_req_t *req, h2o_tunnel_t *tunnel, uint64_t idle
     h2o_iovec_t datagram_flow_id = {};
 
     if (stream->tunnel == NULL) {
+        tunnel->destroy(tunnel);
```

Comment:
This approach would only work if there's a guarantee that the stream does not get freed before `establish_tunnel` is called, even though it _may_ enter CLOSE_WAIT state (at which point `stream->tunnel` is set to NULL).

I do not think that assumption holds. Hence the alternate approach.